tsl/src/remote/dist_ddl.c

```diff
@@ -421,20 +436,155 @@ dist_ddl_process_rename(ProcessUtilityArgs *args)
 		 * Save oid for dist_ddl_end execution.
 		 */
 		dist_ddl_state.relid = linitial_oid(args->hypertable_list);
-		dist_ddl_state_set_exec_type(DIST_DDL_EXEC_ON_END);
+		dist_ddl_state_schedule(DIST_DDL_EXEC_ON_END, args);
 		return;
 	}
 
 	/* Only handles other renamings here (e.g., triggers) */
 	if (dist_ddl_state_set_hypertable(args))
-		dist_ddl_state_set_exec_type(DIST_DDL_EXEC_ON_START);
+		dist_ddl_state_schedule(DIST_DDL_EXEC_ON_START, args);
+}
+
+static char *
+dist_ddl_process_grant_on_database_for(ProcessUtilityArgs *args, const char *dbname)
+{
+	GrantStmt *stmt = castNode(GrantStmt, args->parsetree);
+	ListCell *lc;
+
+	/*
+	   GRANT { { CREATE | CONNECT | TEMPORARY | TEMP } [, ...] | ALL [ PRIVILEGES ] }
+	   ON DATABASE database_name [, ...]
+	   TO role_specification [, ...] [ WITH GRANT OPTION ]
+
+	   REVOKE [ GRANT OPTION FOR ]
+	   { { CREATE | CONNECT | TEMPORARY | TEMP } [, ...] | ALL [ PRIVILEGES ] }
+	   ON DATABASE database_name [, ...]
+	   FROM role_specification [, ...]
+	   [ CASCADE | RESTRICT ]
+	*/
+	StringInfo command = makeStringInfo();
+
+	/* GRANT/REVOKE */
+	if (stmt->is_grant)
+		appendStringInfo(command, "GRANT ");
+	else
+		appendStringInfo(command, "REVOKE ");
+
+	/* privileges [, ...] | ALL */
+	if (stmt->privileges == NULL)
+	{
+		/* ALL */
+		appendStringInfo(command, "ALL ");
+	}
+	else
+	{
+		foreach (lc, stmt->privileges)
+		{
+			AccessPriv *priv = lfirst(lc);
+
+			appendStringInfo(command,
+							 "%s%s ",
+							 priv->priv_name,
+							 lnext(stmt->privileges, lc) != NULL ? "," : "");
+		}
+	}
+
+	/* Set database name of the data node */
+	appendStringInfo(command, "ON DATABASE %s ", dbname);
```

Comment:
Do we have a convenient way to use parameterised statements here? Concatenating user input to build SQL queries always feels uncomfortable... Is it at least quoted here, e.g. if my database is "my-database"?